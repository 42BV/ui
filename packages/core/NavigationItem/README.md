# @42.nl/ui-core-navigation-item

## Installation

Install the package in your project directory with:

```sh
// with npm
npm install @42.nl/ui-core-navigation-item

// with yarn
yarn add @42.nl/ui-core-navigation-item
```
