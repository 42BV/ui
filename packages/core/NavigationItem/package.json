{
  "name": "@42.nl/ui-core-navigation-item",
  "version": "0.0.1",
  "description": "",
  "main": "lib/NavigationItem.js",
  "private": false,
  "scripts": {
    "test": "jest --config=jest.config.js",
    "build": "tsc",
    "lint": "eslint src --ext js,jsx,ts,tsx --max-warnings=0",
    "clean": "rm -rf lib"
  },
  "files": [
    "lib/*"
  ],
  "dependencies": {
    "@42.nl/ui-core-icon": "0.0.1",
    "classnames": "2.2.6"
  },
  "peerDependencies": {
    "react": "^16.9.0",
    "reactstrap": "^8.0.1",
    "react-router-dom": "^5.0.1"
  },
  "devDependencies": {
    "@types/react-router-dom": "4.3.4",
    "react": "16.9.0",
    "reactstrap": "8.0.1",
    "react-router-dom": "5.0.1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "publishConfig": {
    "access": "public"
  }
}
