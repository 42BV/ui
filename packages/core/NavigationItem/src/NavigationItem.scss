@import '~@42.nl/ui-core-styling/scss/bootstrap';
@import '~@42.nl/ui-core-icon/lib/scss/Icon';
