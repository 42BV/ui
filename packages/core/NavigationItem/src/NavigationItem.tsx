import React from 'react';
import { NavLink as RRNavLink } from 'react-router-dom';
import { NavItem, NavLink } from 'reactstrap';

import { Icon, IconType } from '@42.nl/ui-core-icon';
import { useCurrentUser } from '@42.nl/authentication';

interface Props {
  /**
   * String that corresponds to a route.
   *
   * @type {string}
   * @memberof Props
   */
  to: string;

  /**
   * Determines which icon to render above the text.
   *
   * @type {IconType}
   * @memberof Props
   */
  icon: IconType;

  /**
   * Link text as rendered underneath icon.
   *
   * @type {string}
   * @memberof Props
   */
  text?: string;

  /**
   * Guard that that prevents display by user role.
   *
   * @type {string}
   * @memberof Props
   */
  requiresRole?: string;
}

/**
 *
 *
 * @export NavigationItem
 * @param {Props} props
 * @returns {JSX.Element}
 */
export default function NavigationItem({
  requiresRole,
  to,
  icon,
  text
}: Props): JSX.Element {
  const currentUser = useCurrentUser<any>();
  const userRoles = currentUser.roles;

  if (requiresRole !== undefined && !userRoles.includes(requiresRole)) {
    return null;
  }

  return (
    <NavItem>
      <NavLink to={to} exact tag={RRNavLink} activeClassName="active">
        <Icon icon={icon} className="mr-3 align-bottom" />
        {text}
      </NavLink>
    </NavItem>
  );
}
