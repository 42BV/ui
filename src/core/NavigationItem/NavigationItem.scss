.navigation-item {
  list-style: none;
}
