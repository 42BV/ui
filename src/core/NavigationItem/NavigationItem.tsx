import React from 'react';
import { NavLink as RRNavLink } from 'react-router-dom';
import { NavItem, NavLink } from 'reactstrap';
import classNames from 'classnames';

import { Icon, IconType } from '../Icon';

interface Props {
  /**
   * String that corresponds to a route.
   */
  to: string;

  /**
   * Determines which icon to render above the text.
   */
  icon: IconType;

  /**
   * Link text as rendered underneath icon.
   */
  text?: string;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;

  /**
   * Predicate to determine if the link will be shown. Accepts either a boolean or a function that returns a boolean.
   */
  show?: (() => boolean) | boolean;

  /**
   * Optionally whether the route is an exact route.
   * 
   * Defaults to true.
   */
  exact?: boolean;
}

/**
 * The NavigationItem is a wrapper around the `react-router`'s 
 * `NavLink`. See https://reacttraining.com/react-router/web/api/NavLink.
 * It adds texts and icons around the `NavLink`.
 * 
 * The NavigationItem enables you to guard specific links by a predicate, 
 * for example by user role.
 *
 * Use it when you want to keep certain navigation items hidden 
 * for specific user roles.
 * 
 * By default the `NavigationItem` will render an `exact` link. See:
 * https://reacttraining.com/react-router/web/api/Route/exact-bool.
 * But you can set it to `false` via the props.
 */
export default function NavigationItem({
  to,
  icon,
  text,
  show = true,
  className,
  exact = true
}: Props) {
  const shouldShow = typeof show === 'function' ? show : () => show;

  if (!shouldShow()) {
    return null;
  }

  return (
    <NavItem className={classNames('navigation-item', className)}>
      <NavLink to={to} exact={exact} tag={RRNavLink} activeClassName="active">
        <Icon icon={icon} className="mr-3 align-bottom" />
        {text}
      </NavLink>
    </NavItem>
  );
}
