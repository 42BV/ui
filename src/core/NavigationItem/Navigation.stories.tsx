import React from 'react';
import { storiesOf } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

import NavigationItem from './NavigationItem';

storiesOf('core|NavigationItem', module)
  .addParameters({ component: NavigationItem })
  .add('default', () => {
    return (
      <BrowserRouter>
        <NavigationItem to="/dashboard" icon="dashboard" text="Dashboard" />
      </BrowserRouter>
    );
  })
  .add('with show boolean', () => {
    return (
      <BrowserRouter>
        <NavigationItem
          show={true}
          to="/dashboard"
          icon="dashboard"
          text="Dashboard"
        />
      </BrowserRouter>
    );
  })
  .add('with show as function', () => {
    return (
      <BrowserRouter>
        <NavigationItem
          show={() => true && true}
          to="/dashboard"
          icon="dashboard"
          text="Dashboard"
        />
      </BrowserRouter>
    );
  })
  .add('with exact is false', () => {
    return (
      <BrowserRouter>
        <NavigationItem
          to="/dashboard"
          icon="dashboard"
          text="Dashboard"
          exact={false}
        />
      </BrowserRouter>
    );
  });
