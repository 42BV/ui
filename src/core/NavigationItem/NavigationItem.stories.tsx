import React from 'react';
import { storiesOf } from '@storybook/react';
import { BrowserRouter } from 'react-router';

import { NavigationItem } from './NavigationItem';
import { Alert } from 'reactstrap';

storiesOf('core/NavigationItem', module)
  .addParameters({ component: NavigationItem })
  .addDecorator((Story) => (
    <>
      <Alert color="warning" className="mb-4">
        <p>
          To be able to use NavigationItem, you have to add react-router to your
          dependencies:
        </p>
        <code>npm install --save react-router</code>
      </Alert>
      <Story />
    </>
  ))
  .add('default', () => {
    return (
      <BrowserRouter>
        <ul>
          <NavigationItem to="/dashboard" icon="dashboard" text="Dashboard" />
        </ul>
      </BrowserRouter>
    );
  })
  .add('with show boolean', () => {
    return (
      <BrowserRouter>
        <ul>
          <NavigationItem
            show={true}
            to="/dashboard"
            icon="dashboard"
            text="Dashboard"
          />
        </ul>
      </BrowserRouter>
    );
  })
  .add('with show as function', () => {
    return (
      <BrowserRouter>
        <ul>
          <NavigationItem
            show={() => true}
            to="/dashboard"
            icon="dashboard"
            text="Dashboard"
          />
        </ul>
      </BrowserRouter>
    );
  });
