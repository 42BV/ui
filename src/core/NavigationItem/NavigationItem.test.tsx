import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { NavigationItem } from './NavigationItem';
import { Navigator, Router } from 'react-router-dom';

describe('Component: NavigationItem', () => {
  function setup({
    show,
    className,
    location = 'test'
  }: {
    show?: (() => boolean) | boolean;
    className?: string;
    location?: string;
  }) {
    // @ts-expect-error Test mock
    const navigator: Navigator = {
      createHref: (to: string) => to
    };

    const { container } = render(
      <Router location={location} navigator={navigator}>
        <NavigationItem
          to="/dashboard"
          icon="dashboard"
          text="Dashboard"
          show={show}
          className={className}
        />
      </Router>
    );

    return { container };
  }

  describe('ui', () => {
    test('default', () => {
      const { container } = setup({});
      expect(container).toMatchSnapshot();
    });

    test('active', () => {
      setup({ location: '/dashboard' });
      expect(screen.getByText('dashboard').parentNode).toHaveClass('active');
    });

    test('extra class', () => {
      const { container } = setup({ className: 'extra-class' });
      expect(container.firstChild).toHaveClass('extra-class');
    });
  });

  describe('show behavior', () => {
    it('should not render when show is false', () => {
      setup({ show: false });
      expect(screen.queryByText('dashboard')).not.toBeInTheDocument();
    });

    it('should render when predicate resolves to true', () => {
      setup({ show: () => 1 + 1 === 2 });
      expect(screen.queryByText('dashboard')).toBeInTheDocument();
    });

    it('should not render when predicate resolves to false', () => {
      setup({ show: () => 1 + 2 === 2 });
      expect(screen.queryByText('dashboard')).not.toBeInTheDocument();
    });

    it('should default show to true if it is not provided', () => {
      setup({});
      expect(screen.queryByText('dashboard')).toBeInTheDocument();
    });
  });
});
