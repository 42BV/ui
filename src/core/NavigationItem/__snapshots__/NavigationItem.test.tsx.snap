// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: NavigationItem ui default 1`] = `
<div>
  <li
    class="navigation-item nav-item"
  >
    <a
      class="nav-link"
      data-discover="true"
      href="[object Object]"
    >
      <i
        class="icon material-icons material-icons__filled me-3 align-bottom"
      >
        dashboard
      </i>
      Dashboard
    </a>
  </li>
</div>
`;
